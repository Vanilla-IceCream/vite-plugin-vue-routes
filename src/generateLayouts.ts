import path from 'path';
import { glob } from 'glob';

export default async () => {
  const layouts = await glob(`src/layouts/*.vue`);

  const asyncLayouts = layouts.map((layout) => {
    const importName = path.basename(layout, '.vue');
    const importPath = path.resolve(process.cwd(), layout);
    return `${importName}: defineAsyncComponent(() => import('${importPath}')),`;
  });

  return `
    <script setup>
    import { defineAsyncComponent, ref, watch, markRaw } from 'vue';
    import { useRoute } from 'vue-router';

    const layout = ref();
    const route = useRoute();

    const layouts = { ${asyncLayouts.join('')} };

    watch(
      () => route.meta?.layout,
      (val) => {
        layout.value = markRaw(layouts[val || 'Default']);
      },
      { immediate: true },
    );
    </script>

    <template>
      <component :is="layout"><slot></slot></component>
    </template>
  `;
};
