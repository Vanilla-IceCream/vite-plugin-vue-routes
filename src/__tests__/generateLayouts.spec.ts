import { test, expect } from 'vitest';

import generateLayouts from '../generateLayouts';

test('generateLayouts', async () => {
  const generated = await generateLayouts();
  expect(generated).toMatchSnapshot();
});

test('generateLayouts - sfc', async () => {
  const generated = await generateLayouts(true);
  expect(generated).toMatchSnapshot();
});
