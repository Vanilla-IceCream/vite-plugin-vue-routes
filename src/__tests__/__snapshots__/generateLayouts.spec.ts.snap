// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateLayouts - sfc 1`] = `
"
      <script setup>
      import { defineComponent, defineAsyncComponent, ref, watch, markRaw, useSlots, h } from 'vue';
      import { useRoute } from 'vue-router';

      const layout = ref();
      const slots = useSlots();
      const route = useRoute();

      const layouts = {
        __Placeholder: defineComponent(() => () => h('div', slots.default())),
        
      };

      watch(
        () => route.meta?.layout,
        (val) => {
          layout.value = markRaw(layouts[val || '__Placeholder']);
        },
        { immediate: true },
      );
      </script>

      <template>
        <component :is=\\"layout\\"><slot></slot></component>
      </template>
    "
`;

exports[`generateLayouts 1`] = `
"
    import { defineComponent, defineAsyncComponent, ref, watch, markRaw, useSlots, h } from 'vue';
    import { useRoute } from 'vue-router';

    export default defineComponent({
      setup() {
        const layout = ref();
        const slots = useSlots();
        const route = useRoute();

        const layouts = {
          __Placeholder: defineComponent(() => () => h('div', slots.default())),
          
        };

        watch(
          () => route.meta?.layout,
          (val) => {
            layout.value = markRaw(layouts[val || '__Placeholder']);
          },
          { immediate: true },
        );

        return {
          layout,
        };
      },
      render() {
        return h(this.layout, {}, () => this.$slots.default());
      },
    });
  "
`;
